homeassistant/helpers/trigger.py

```diff
@@ -621,3 +625,107 @@ async def async_get_all_descriptions(
+            if (
+                event.data["entity_id"] in selected.referenced
+                or event.data["entity_id"] in selected.indirectly_referenced
+            ):
+                self._action(event)
+
+        tracked_entities = selected.referenced.union(selected.indirectly_referenced)
+
+        _LOGGER.debug("Tracking state changes for entities: %s", tracked_entities)
+        return async_track_state_change_event(
+            self._hass, tracked_entities, state_change_listener, job_type=self._job_type
+        )
+
+    def _setup_tracking(self) -> None:
+        """Initialize state change tracking."""
+        self._state_change_unsub = self._track_entities_state_change()
+
+    def _setup_registry_listeners(self) -> None:
+        """Set up listeners for registry changes that require resubscription."""
+
+        @callback
+        def resubscribe_state_change_event(event: Event[Any] | None = None) -> None:
+            """Resubscribe to state change events when registry changes."""
+            if self._state_change_unsub:
+                self._state_change_unsub()
+            self._state_change_unsub = self._track_entities_state_change()
+
+        self._registry_unsubs = [
+            self._hass.bus.async_listen(
+                entity_registry.EVENT_ENTITY_REGISTRY_UPDATED,
+                resubscribe_state_change_event,
+                # TODO(abmantis): filter for entities that match the target selector?
+                # event_filter=self._filter_entity_registry_changes,
+            ),
+            self._hass.bus.async_listen(
+                device_registry.EVENT_DEVICE_REGISTRY_UPDATED,
+                resubscribe_state_change_event,
+            ),
+            self._hass.bus.async_listen(
+                area_registry.EVENT_AREA_REGISTRY_UPDATED,
+                resubscribe_state_change_event,
+            ),
+        ]
+
+    def unsub(self) -> None:
+        """Unsubscribe from all events."""
+        for registry_unsub in self._registry_unsubs:
+            registry_unsub()
+        self._registry_unsubs.clear()
+        if self._state_change_unsub:
+            self._state_change_unsub()
+            self._state_change_unsub = None
+
+
+def async_track_target_selector_state_change_event(
+    hass: HomeAssistant,
+    target_selector_config: ConfigType,
+    action: Callable[[Event[EventStateChangedData]], Any],
+    job_type: HassJobType | None = None,
+) -> CALLBACK_TYPE:
+    """Track state changes for entities referenced directly or indirectly (by device, area, label, etc) in a target selector."""
+    selector_data = TargetSelectorData(target_selector_config)
+    if not selector_data.has_any_selector:
+        _LOGGER.warning(
+            "Target selector %s does not have any selectors defined",
+            target_selector_config,
+        )
+        return lambda: None
+
+    tracker = TargetSelectorStateChangeTracker(hass, selector_data, job_type, action)
```

Comment:
Initially I had a functional implementation, but I wasn't a big fan of it so I moved it into a class. I think it is easier to read now.

The functional version is here: [`5d553e5` (#148086)](https://github.com/home-assistant/core/pull/148086/commits/5d553e56411315b676e0e43ca5f14fd47e451dd4)

Reply:
I'm not sure this function adds value, why don't users just create a `TargetSelectorStateChangeTracker` themselves?